File: src/main/java/com/xemantic/test/howfast/XORShift.java
/*
 * Copyright 2021  Sam Leonard
 *
 * This file is part of java-2-times-faster-than-c.
 *
 * java-2-times-faster-than-c is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * java-2-times-faster-than-c is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with shader-web-background.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.xemantic.test.howfast;

class xorshift64s_rng {
	private long state;

	xorshift64s_rng(long state) {
		this.state = state;
	}

	double get_rand() {
		long x = this.state;	/* The state must be seeded with a nonzero value. */
		x ^= x >>> 12; // a
		x ^= x  << 25; // b
		x ^= x >>> 27; // c
		this.state = x;

		long rand_val = x * 0x2545F4914F6CDD1DL;

		// mix to a double
		long a = (rand_val >>> 32) & 0xFFFFFFFFL;
		long b = rand_val & 0xFFFFFFFFL;

		return ((a >> 5) * 67108864.0 + (b >> 6)) * (1.0 / 9007199254740991.0);
	}
}

public class XORShift {

  private static final long ITERATION_COUNT = 1000000000L;

  public static void main(String[] args) {
		xorshift64s_rng rng = new xorshift64s_rng(42);

    double checksum = 0;
    for (long i = 0; i < ITERATION_COUNT; i++) {
      checksum += rng.get_rand();
    }
    System.out.println("checksum: " + checksum);
  }

}

File: src/main/kotlin/XorShift.kt
/*
 * Copyright 2021  Kazimierz Pogoda
 *
 * This file is part of java-2-times-faster-than-c.
 *
 * java-2-times-faster-than-c is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * java-2-times-faster-than-c is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with shader-web-background.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.xemantic.test.howfast.kotlin

import kotlin.math.sin

const val ITERATION_COUNT = 1000000000L

@ExperimentalUnsignedTypes
class XorShift64SRng(var state: ULong) {
  init {
    assert(state != 0UL)
  }

  fun getRand(): Double {
    var x: ULong = state /* The state must be seeded with a nonzero value. */
    x = x xor (x shr 12) // a
    x = x xor (x shl 25) // b
    x = x xor (x shr 27) // c

    state = x
    val rand_val: ULong = x * 0x2545F4914F6CDD1DUL;

    // mix to a double
    val a = (rand_val shr 32).toUInt()
    val b = (rand_val and 0xFFFFFFFFU).toUInt()

    return ((a shr 5).toDouble() * 67108864.0 + (b shr 6).toDouble()) * (1.0 / 9007199254740991.0)
  }
}

fun main() {
  var rng = XorShift64SRng(42UL);
  var checksum = 0.0
  for (i in 0 until ITERATION_COUNT) {
    checksum += rng.getRand()
  }
  println("checksum: $checksum")
}
